src/inference/ais.jl

```diff
@@ -0,0 +1,339 @@
+    # use path VarInfo to build instance of AISTransition
+    nt = NamedTuple()
+    theta = merge(DynamicPPL.tonamedtuple(spl.state.vi), NamedTuple())
+    return AISTransition(theta, lp, accum_logweight)
+end
+
+# B.4. sample_end! combines the individual accum_logweights to obtain final_logevidence, as in vanilla IS 
+
+"""
+    AbstractMCMC.sample_end!( ::AbstractRNG, ::Model, spl::Sampler{<:AIS}, N::Integer, ts::Vector; kwargs...)
+
+Store estimate of the log evidence in the AISState attribute final_logevidence of spl.state.
+"""
+function AbstractMCMC.sample_end!(
+    ::AbstractRNG,
+    ::Model,
+    spl::Sampler{<:AIS},
+    N::Integer,
+    ts::Vector;
+    kwargs...
+)
+    # use AISTransition accum_logweight attribute to store estimate of log evidence
+    spl.state.final_logevidence = logsumexp(map(x->x.accum_logweight, ts)) - log(N)
+end
+
+# C. helper functions
+
+
+"""
+    gen_logjoint(v, model, spl)
+
+Return the log joint density function corresponding to model.
+"""
+function gen_logjoint(v, model, spl)
+    function logjoint(z)::Float64
+        z_old, lj_old = v[spl], getlogp(v) 
+        v[spl] = z
+        model(v, spl)
+        lj = getlogp(v)
+        v[spl] = z_old
+        setlogp!(v, lj_old)
+        return lj
+    end
+    return logjoint
+end
+
+"""
+    gen_logprior(v, model, spl)
+
+Return the log prior density function corresponding to model.
+"""
+function gen_logprior(v, model, spl)
+    function logprior(z)::Float64
+        z_old, lj_old = v[spl], getlogp(v)
+        v[spl] = z
+        model(v, SampleFromPrior(), PriorContext())
+        lj = getlogp(v)
+        v[spl] = z_old
+        setlogp!(v, lj_old)
+        return lj
+    end
+    return logprior
+end
+
+"""
+    gen_log_unnorm_tempered(logprior, logjoint, beta)
+
+Return the log unnormalized tempered density function corresponding to model, ie a convex combination of the logprior and logjoint densities with parameter beta.
+"""
+function gen_log_unnorm_tempered(logprior, logjoint, beta)
+    function log_unnorm_tempered(z)
+        return (1 - beta) * logprior(z) + beta * logjoint(z)
+    end
+    return log_unnorm_tempered
+end
+
+"""
+    prior_step(spl, model)
+
+Sample from prior to return inital values of current_state and accum_logweight.
+"""
+function prior_step(spl, model)
+    # sample from prior
+    prior_vi = VarInfo()
+    prior_spl = SampleFromPrior()
+    model(prior_vi, prior_spl)
+
+    # initialize current_state
+    current_state = prior_vi[prior_spl]
+
+    # initialize accum_logweight
+    logprior = gen_logprior(spl.state.vi, model, spl)
+    accum_logweight = - logprior(current_state)
+
+    return current_state, accum_logweight
+end
+
+"""
+    intermediate_step(j, spl, current_state, accum_logweight)
+
+Perform the MCMC step corresponding to the j-th intermediate distribution, with the j-th MH proposal. Return updated current_state and accum_logweight.
+"""
+function intermediate_step(j, spl, current_state, accum_logweight)
```

Comment:
```suggestion
function intermediate_step(densitymodel, proposal_kernel, current_state, accum_logweight)
```
And then remove the def of `densitymodel` and `proposal_kernel` below, since `spl` is never used in this method.